### ISSUE TITLE
Allowing async calls in ClientHooks.onRequestPrepared

### ISSUE THREAD

**Issue by b099l3** [syncapod/tart#7]
# Context
I want to use hooks to add auth headers to the request. To get the auth token I am using an async method (reading from secure storage) and then adding it to the headers. However Hooks only has synchronous methods and requires a method signature that returns `Context`.

# Proposed Solution
 Add an async version of the `onRequestPrepared` property on `ClientHooks`, i.e.,
``` dart
/// onRequestPrepared is after request has been prepared and immediately before request is sent
Future<Context> Function(Context ctx, http.Request req) onRequestPrepared;
```
  also changed the default method to:
```dart
static Future<Context> defaultOnRequestPrepared(ctx, req) => ctx;
```
  
 It would mean a change in the generator `doProtobufRequest` and `doJSONRequest` methods to await the call, so adding
``` dart
// call onRequestPrepared hook for user to modify request
ctx = await hooks.onRequestPrepared(ctx, req);
```

# What the code would look like
  
This then allows for code like this:
  
``` dart
ProtoClient(
  EnvConstants.baseUrl,
  EnvConstants.twirpPath,
  hooks: ClientHooks(
    onRequestPrepared: (ctx, req) async {
      final auth = await ApiClientConfig.authHeader();
      req.headers.addAll(auth);
      return ctx;
    },
  ),
);
```

if there are no async methods called you would still have to add the `async` keyword on the function like this:
  
``` dart
ProtoClient(
  EnvConstants.baseUrl,
  EnvConstants.twirpPath,
  hooks: ClientHooks(
    onRequestPrepared: (ctx, req) async {
      final auth = {'auth' : 'Value'};
      req.headers.addAll(auth);
      return ctx;
    },
  ),
);
```

I have already coded this, so could open a PR but wanted to discuss if this was something you wanted in this library?

**Comment by b099l3** [comment]
It could also be done for `onResponseReceived` and `onError` for `ClientHooks` and because they return void I dont think we would need the `async` on the anonymous functions

**Comment by sschwartz96** [comment]
Hi @b099l3, I apologize for the delay, I meant to respond a couple of days ago.

I agree with your point, this would provide more power to the user in what they are able to do in the onRequestPrepared function.

You bring up another good point that we also need to include awaits in the generated code for when onResponseReceived and onError are called. This will ensure that even if the user passes an async function order it kept when executing the hooks.

Thank you!

**Comment by b099l3** [comment]
I am sorry for how long this took me to come back to. I hope this PR #8 covers the ideas we discussed. Let me know if you want other changes made.

**Comment by sschwartz96** [comment]
Closed with #8

**Comment by sschwartz96** [comment]
@b099l3 thank you for reminding me and for the contribution!
